File: crates/sealed-lattice-kernel/src/bgv/setup/tests/accepted_setup/package_fixtures/common_randomness.rs
use super::*;

use crate::hashing::derive_canonical_object_hash;

#[allow(clippy::too_many_arguments)]
pub(super) fn common_randomness_object(
    ceremony_id: &str,
    manifest_hash: &str,
    roster_hash: &str,
    setup_parameters_hash: &str,
    setup_epoch: &str,
    participant_count: u64,
) -> serde_json::Value {
    let setup_context_hash =
        crate::bgv::setup::accepted_setup::setup_context_hash(&serde_json::json!({
            "ceremonyId": ceremony_id,
            "manifestHash": manifest_hash,
            "rosterHash": roster_hash,
            "setupParametersHash": setup_parameters_hash,
            "setupEpoch": setup_epoch,
            "participantCount": participant_count,
        }))
        .expect("setup context hash");
    let mut commit_records = Vec::new();
    let mut reveal_records = Vec::new();
    let mut ordered_reveal_hashes = Vec::new();
    for roster_position in 0..participant_count {
        let trustee_identity = format!("trustee-{roster_position}");
        let reveal_source_hash = derive_canonical_object_hash(&serde_json::json!({
            "objectType": "CommonRandomnessRevealHash",
            "fixture": "common-randomness-reveal",
            "rosterPosition": roster_position,
        }))
        .expect("reveal source hash");
        let reveal_hex = reveal_source_hash[..64].to_string();
        let reveal_payload = serde_json::json!({
            "objectType": "CommonRandomnessReveal",
            "setupContextHash": &setup_context_hash,
            "trusteeIdentity": trustee_identity.clone(),
            "rosterPosition": roster_position,
            "recoveryEpoch": 0,
            "deviceEpoch": 0,
            "revealHex": reveal_hex,
        });
        let reveal_hash = derive_canonical_object_hash(&reveal_payload).expect("reveal hash");
        let reveal_context_hash = derive_canonical_object_hash(&serde_json::json!({
            "objectType": "CommonRandomnessRevealSignatureContext",
            "payloadRoot": reveal_hash.as_str(),
        }))
        .expect("reveal signature context hash");
        let signature_seed_label = setup_trustee_signature_seed_label(&trustee_identity);
        let reveal_signature_fixture = create_protocol_signature_fixture(
            &signature_seed_label,
            serde_json::json!({
                "objectType": "CommonRandomnessReveal",
                "ceremonyId": ceremony_id,
                "manifestHash": manifest_hash,
                "objectRoot": reveal_hash.as_str(),
                "signerRole": "Trustee",
                "signerIdentity": trustee_identity.as_str(),
                "recoveryEpoch": 0,
                "deviceEpoch": 0,
                "contextHash": reveal_context_hash,
            }),
        )
        .expect("reveal signature fixture");
        ordered_reveal_hashes.push(reveal_hash.clone());
        reveal_records.push(serde_json::json!({
            "objectType": "CommonRandomnessReveal",
            "rosterPosition": roster_position,
            "revealHex": reveal_hex,
            "signatureEnvelope": reveal_signature_fixture.envelope,
        }));

        let commit_payload = serde_json::json!({
            "objectType": "CommonRandomnessCommit",
            "setupContextHash": &setup_context_hash,
            "trusteeIdentity": trustee_identity.as_str(),
            "rosterPosition": roster_position,
            "recoveryEpoch": 0,
            "deviceEpoch": 0,
            "revealHash": reveal_hash.as_str(),
        });
        let commit_hash = derive_canonical_object_hash(&commit_payload).expect("commit hash");
        let commit_context_hash = derive_canonical_object_hash(&serde_json::json!({
            "objectType": "CommonRandomnessCommitSignatureContext",
            "payloadRoot": commit_hash.as_str(),
        }))
        .expect("commit signature context hash");
        let commit_signature_fixture = create_protocol_signature_fixture(
            &signature_seed_label,
            serde_json::json!({
                "objectType": "CommonRandomnessCommit",
                "ceremonyId": ceremony_id,
                "manifestHash": manifest_hash,
                "objectRoot": commit_hash.as_str(),
                "signerRole": "Trustee",
                "signerIdentity": trustee_identity.as_str(),
                "recoveryEpoch": 0,
                "deviceEpoch": 0,
                "contextHash": commit_context_hash,
            }),
        )
        .expect("commit signature fixture");
        commit_records.push(serde_json::json!({
            "objectType": "CommonRandomnessCommit",
            "rosterPosition": roster_position,
            "revealHash": reveal_hash,
            "signatureEnvelope": commit_signature_fixture.envelope,
        }));
    }

    let public_matrix_seed_hash = derive_canonical_object_hash(&serde_json::json!({
        "objectType": "SetupPublicMatrixSeed",
        "setupContextHash": setup_context_hash,
        "orderedRevealHashes": ordered_reveal_hashes,
    }))
    .expect("public matrix seed hash");
    serde_json::json!({
        "objectType": "SetupCommonRandomness",
        "commitRecords": commit_records,
        "revealRecords": reveal_records,
        "publicMatrixSeedHash": public_matrix_seed_hash,
    })
}

File: packages/wasm/tests/node/transcript-core-kernel/bgv-collective-setup/accepted-setup-package-fixtures/common-randomness.ts
import { expect } from 'vitest';

import { setupRequest } from '../../bgv-passive-setup-fixtures.js';
import {
    collectiveSetupRosterHash,
    setupTrusteeSignatureSeedLabel,
    textEncoder,
    type JsonRecord,
} from '../setup-fixture-primitives.js';

import { canonicalJson } from '#packages/crypto/src/index';
import {
    createMlDsaKeyPairFixture,
    createMlDsaSignatureProfileFixture,
    createProtocolSignatureFixture,
} from '#packages/crypto/tests/support/protocol-signature-fixtures';
import type {
    BgvCollectiveSetupParametersDescription,
    TranscriptCoreKernel,
} from '#packages/wasm/src/index';

const canonicalByteLength = (value: unknown): number =>
    textEncoder.encode(canonicalJson(value)).byteLength;

function commonRandomnessSignatureContextHash(input: {
    readonly kernel: TranscriptCoreKernel;
    readonly objectType: 'CommonRandomnessCommit' | 'CommonRandomnessReveal';
    readonly purpose:
        | 'common-randomness-commit-signature-context'
        | 'common-randomness-reveal-signature-context';
    readonly payload: JsonRecord;
    readonly objectRoot: string;
}): string {
    return input.kernel.deriveCanonicalObjectHash({
        value: {
            objectType: `${input.objectType}SignatureContext`,
            purpose: input.purpose,
            ceremonyId: input.payload.ceremonyId,
            manifestHash: input.payload.manifestHash,
            rosterHash: input.payload.rosterHash,
            setupParametersHash: input.payload.setupParametersHash,
            setupEpoch: input.payload.setupEpoch,
            trusteeIdentity: input.payload.trusteeIdentity,
            rosterPosition: input.payload.rosterPosition,
            objectRoot: input.objectRoot,
        },
    });
}

function commonRandomnessSignatureEnvelope(input: {
    readonly kernel: TranscriptCoreKernel;
    readonly objectType: 'CommonRandomnessCommit' | 'CommonRandomnessReveal';
    readonly purpose:
        | 'common-randomness-commit-signature-context'
        | 'common-randomness-reveal-signature-context';
    readonly payload: JsonRecord;
    readonly objectRoot: string;
    readonly trusteeIdentity: string;
}): JsonRecord {
    const keyFixture = createMlDsaKeyPairFixture(
        setupTrusteeSignatureSeedLabel(input.trusteeIdentity),
    );
    const signatureContextHash = commonRandomnessSignatureContextHash(input);

    return createProtocolSignatureFixture({
        profile: createMlDsaSignatureProfileFixture(),
        publicKeyBytesHex: keyFixture.publicKeyBytesHex,
        publicKeyHash: keyFixture.publicKeyHash,
        secretKeyBytesHex: keyFixture.secretKeyBytesHex,
        signedRoot: {
            objectType: input.objectType,
            ceremonyId: String(input.payload.ceremonyId),
            manifestHash: String(input.payload.manifestHash),
            boardHeadHash: null,
            objectRoot: input.objectRoot,
            chunkMerkleRoot: null,
            byteLength: canonicalByteLength(input.payload),
            signerRole: 'Trustee',
            signerIdentity: input.trusteeIdentity,
            recoveryEpoch: Number(input.payload.recoveryEpoch),
            deviceEpoch: Number(input.payload.deviceEpoch),
            contextHash: signatureContextHash,
        },
    });
}

export function acceptedCommonRandomness(
    kernel: TranscriptCoreKernel,
    setupParameters: BgvCollectiveSetupParametersDescription,
): JsonRecord {
    const commitRecords: JsonRecord[] = [];
    const revealRecords: JsonRecord[] = [];
    const orderedRevealHashes: string[] = [];
    const rosterHash = collectiveSetupRosterHash(
        (input) => kernel.deriveCanonicalObjectHash(input),
        setupParameters.participantCount,
    );
    for (
        let rosterPosition = 0;
        rosterPosition < setupParameters.participantCount;
        rosterPosition += 1
    ) {
        const trusteeIdentity = `trustee-${String(rosterPosition)}`;
        const revealHex = kernel
            .deriveCanonicalObjectHash({
                value: {
                    objectType: 'CommonRandomnessRevealHash',
                    fixture: 'common-randomness-reveal',
                    rosterPosition,
                },
            })
            .slice(0, 64);
        const revealPayload: JsonRecord = {
            objectType: 'CommonRandomnessReveal',
            ceremonyId: setupRequest.ceremonyId,
            manifestHash: setupRequest.manifestHash,
            rosterHash,
            setupParametersHash: setupParameters.setupParametersHash,
            setupEpoch: 'setup-epoch-1',
            signerRole: 'Trustee',
            trusteeIdentity,
            rosterPosition,
            recoveryEpoch: 0,
            deviceEpoch: 0,
            revealHex,
        };
        const revealHash = kernel.deriveCanonicalObjectHash({
            value: revealPayload,
        });
        const revealSignatureEnvelope = commonRandomnessSignatureEnvelope({
            kernel,
            objectType: 'CommonRandomnessReveal',
            purpose: 'common-randomness-reveal-signature-context',
            payload: revealPayload,
            objectRoot: revealHash,
            trusteeIdentity,
        });
        const revealRecord: JsonRecord = {
            ...revealPayload,
            signatureEnvelopeHash: revealSignatureEnvelope.signatureHash,
            signatureEnvelope: revealSignatureEnvelope,
            revealHash,
        };
        revealRecords.push(revealRecord);
        orderedRevealHashes.push(revealHash);

        const commitPayload: JsonRecord = {
            objectType: 'CommonRandomnessCommit',
            ceremonyId: setupRequest.ceremonyId,
            manifestHash: setupRequest.manifestHash,
            rosterHash,
            setupParametersHash: setupParameters.setupParametersHash,
            setupEpoch: 'setup-epoch-1',
            signerRole: 'Trustee',
            trusteeIdentity,
            rosterPosition,
            recoveryEpoch: 0,
            deviceEpoch: 0,
            revealHash,
        };
        const commitHash = kernel.deriveCanonicalObjectHash({
            value: commitPayload,
        });
        const commitSignatureEnvelope = commonRandomnessSignatureEnvelope({
            kernel,
            objectType: 'CommonRandomnessCommit',
            purpose: 'common-randomness-commit-signature-context',
            payload: commitPayload,
            objectRoot: commitHash,
            trusteeIdentity,
        });
        const commitRecord: JsonRecord = {
            ...commitPayload,
            signatureEnvelopeHash: commitSignatureEnvelope.signatureHash,
            signatureEnvelope: commitSignatureEnvelope,
            commitHash,
        };
        commitRecords.push(commitRecord);
    }

    const publicMatrixSeedHash = kernel.deriveCanonicalObjectHash({
        value: {
            objectType: 'SetupPublicMatrixSeed',
            ceremonyId: setupRequest.ceremonyId,
            manifestHash: setupRequest.manifestHash,
            rosterHash,
            setupParametersHash: String(setupParameters.setupParametersHash),
            setupEpoch: 'setup-epoch-1',
            orderedRevealHashes,
        },
    });
    const publicDerivations = kernel.deriveCollectiveBgvSetupPublicDerivations({
        publicMatrixSeedHash,
        decryptionThreshold: setupParameters.qDec,
    });
    expect(
        publicDerivations.publicMatrices.commitmentMatrix.sampledEntries[0]
            ?.coefficientValue,
    ).toEqual(expect.any(Number));
    const commonRandomness: JsonRecord = {
        objectType: 'SetupCommonRandomness',
        ceremonyId: setupRequest.ceremonyId,
        manifestHash: setupRequest.manifestHash,
        rosterHash,
        setupParametersHash: setupParameters.setupParametersHash,
        setupEpoch: 'setup-epoch-1',
        commitRecords,
        revealRecords,
        publicMatrixSeedHash,
        publicDerivations,
    };
    commonRandomness.commonRandomnessRoot = kernel.deriveCanonicalObjectHash({
        value: commonRandomness,
    });

    return commonRandomness;
}

function publicPrivateVssEnvelopeCommitmentReference(
    envelopeReference: JsonRecord,
): JsonRecord {
    const {
        encryptedEnvelope: encryptedEnvelopeForRecipientTransport,
        transportedPrivateVssShareProofMaterial:
            transportedPrivateVssShareProofMaterialForRecipientTransport,
        ...publicReference
    } = envelopeReference;
    void encryptedEnvelopeForRecipientTransport;
    void transportedPrivateVssShareProofMaterialForRecipientTransport;

    return publicReference;
}

export { publicPrivateVssEnvelopeCommitmentReference };

export function publicPrivateVssEnvelopeCommitmentSet(
    privateVssEnvelopeCommitments: JsonRecord,
): JsonRecord {
    return {
        ...privateVssEnvelopeCommitments,
        envelopeReferences: (
            privateVssEnvelopeCommitments.envelopeReferences as JsonRecord[]
        ).map(publicPrivateVssEnvelopeCommitmentReference),
    };
}
